New resource r/aws_wafv2_web_acl (#12688)
terraform-providers/terraform-provider-aws: website/docs/r/wafv2_web_acl.html.markdown

---
subcategory: "WAFv2"
layout: "aws"
page_title: "AWS: aws_wafv2_web_acl"
description: |-
  Creates a WAFv2 Web ACL resource.
---

> # Resource: aws_wafv2_rule_group

Reviewer: ```suggestion
# Resource: aws_wafv2_web_acl
```

Ahh just caught this one when running the website docs locally @pvanbuijtene

    sampled_requests_enabled   = false
  }
}

resource "aws_wafv2_web_acl" "test" {
  name  = "rule-group-example"
  scope = "REGIONAL"

  default_action {
    block {}
  }

  rule {
    name     = "rule-1"
    priority = 1

    override_action {
      count {}
    }

    statement {
      rule_group_reference_statement {
        arn = aws_wafv2_rule_group.example.arn

        excluded_rule {
          name = "rule-to-exclude-b"
        }

        excluded_rule {
          name = "rule-to-exclude-a"
        }
      }
    }

    visibility_config {
      cloudwatch_metrics_enabled = false
      metric_name                = "friendly-rule-metric-name"
      sampled_requests_enabled   = false
    }
  }

  tags = {
    Tag1 = "Value1"
    Tag2 = "Value2"
  }

  visibility_config {
    cloudwatch_metrics_enabled = false
    metric_name                = "friendly-metric-name"
    sampled_requests_enabled   = false
  }
}
```

## Argument Reference

The following arguments are supported:

* `default_action` - (Required) The action to perform if none of the `rules` contained in the WebACL match. See [Default Action](#default-action) below for details.

> * `description` - (Optional) A friendly description of the rule group.

Reviewer: ```suggestion
* `description` - (Optional) A friendly description of WebACL.
```

> * `name` - (Required) A friendly name of the rule group.

Reviewer: ```suggestion
* `name` - (Required) A friendly name of the WebACL.
```

The `field_to_match` block supports the following arguments:

* `all_query_arguments` - (Optional) Inspect all query arguments.
* `body` - (Optional) Inspect the request body, which immediately follows the request headers.
* `method` - (Optional) Inspect the HTTP method. The method indicates the type of operation that the request is asking the origin to perform.
* `query_string` - (Optional) Inspect the query string. This is the part of a URL that appears after a `?` character, if any.
* `single_header` - (Optional) Inspect a single header. See [Single Header](#single-header) below for details.
* `single_query_argument` - (Optional) Inspect a single query argument. See [Single Query Argument](#single-query-argument) below for details.
* `uri_path` - (Optional) Inspect the request URI path. This is the part of a web request that identifies a resource, for example, `/images/daily-ad.jpg`.

### Single Header

> Inspect a single header. Provide the name of the header to inspect, for example, `User-Agent` or `Referer`. This setting isn't case sensitive.

Reviewer: these examples could be misleading to a reader w/o further reading of the bullet points below - to err on the side of caution, we could append parens with something like: "for example, `User-Agent` or `Referer`(provided as lowercase strings)."

The `single_header` block supports the following arguments:

> * `name` - (Optional) The name of the query header to inspect. This setting isn't case sensitive, only use lower case characters.

Reviewer: this line may be a bit contradictory (i.e. for `name` to not be case sensitive but only lower case). we can re-word this to say something like "This setting must be provided as lower case characters.")

The `single_query_argument` block supports the following arguments:

* `name` - (Optional) The name of the query header to inspect. This setting must be provided as lower case characters.

### Text Transformation

The `text_transformation` block supports the following arguments:

* `priority` - (Required) The relative processing order for multiple transformations that are defined for a rule statement. AWS WAF processes all transformations, from lowest priority to highest, before inspecting the transformed content.
* `type` - (Required) The transformation to apply, you can specify the following types: `NONE`, `COMPRESS_WHITE_SPACE`, `HTML_ENTITY_DECODE`, `LOWERCASE`, `CMD_LINE`, `URL_DECODE`. See the [documentation](https://docs.aws.amazon.com/waf/latest/APIReference/API_TextTransformation.html) for more details.

### Visibility Configuration

The `visibility_config` block supports the following arguments:

* `cloudwatch_metrics_enabled` - (Required) A boolean indicating whether the associated resource sends metrics to CloudWatch. For the list of available metrics, see [AWS WAF Metrics](https://docs.aws.amazon.com/waf/latest/developerguide/monitoring-cloudwatch.html#waf-metrics).

> * `metric_name` - (Required) A friendly name of the CloudWatch metric. The name can contain only alphanumeric characters (A-Z, a-z, 0-9) hyphen(-) and underscore (_), with length from one to 128 characters. It can't contain whitespace or metric names reserved for AWS WAF, for example `All` and `Default_Action`.

Reviewer 2: Some Markdown parsers don't special-case a single underscore character, so italicize the remainder of the document. This can be escaped by using `\_` instead

* `sampled_requests_enabled` - (Required) A boolean indicating whether AWS WAF should store a sampling of the web requests that match the rules. You can view the sampled requests through the AWS WAF console.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `arn` - The ARN of the WAF rule group.

Reviewer: ```suggestion
* `arn` - The ARN of the WAF WebACL.
```

* `capacity` - The web ACL capacity units (WCUs) currently being used by this web ACL.

> * `id` - The ID of the WAF rule group.

Reviewer: ```suggestion
* `id` - The ID of the WAF WebACL.
```

## Import

> WAFv2 Rule Group can be imported using the ID, Name and Scope e.g.

Reviewer: nice already resolved 😄
